modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/task/CollectPartitionStatistics.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.ignite.internal.processors.cache.distributed.dht.topology.GridDhtPartitionTopology;
+import org.apache.ignite.internal.processors.cache.persistence.CacheDataRow;
+import org.apache.ignite.internal.processors.query.GridQueryTypeDescriptor;
+import org.apache.ignite.internal.processors.query.h2.opt.GridH2Table;
+import org.apache.ignite.internal.processors.query.h2.opt.H2Row;
+import org.apache.ignite.internal.processors.query.stat.ColumnStatistics;
+import org.apache.ignite.internal.processors.query.stat.ColumnStatisticsCollector;
+import org.apache.ignite.internal.processors.query.stat.GatherStatisticCancelException;
+import org.apache.ignite.internal.processors.query.stat.GatherStatisticRetryException;
+import org.apache.ignite.internal.processors.query.stat.LocalStatisticsGatheringContext;
+import org.apache.ignite.internal.processors.query.stat.ObjectPartitionStatisticsImpl;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.gridgain.internal.h2.table.Column;
+
+import static org.apache.ignite.internal.processors.cache.distributed.dht.topology.GridDhtPartitionState.OWNING;
+
+/**
+ * Implementation of statistic collector.
+ */
+public class CollectPartitionStatistics implements Callable<ObjectPartitionStatisticsImpl> {
+    /** Canceled check interval. */
+    private static final int CANCELLED_CHECK_INTERVAL = 100;
+
+    /** */
+    private final GridH2Table tbl;
+
+    /** */
+    private final Column[] cols;
+
+    /** */
+    private final int partId;
+
+    /** */
+    private final long ver;
+
+    /** */
+    private final Supplier<Boolean> cancelled;
+
+    /** */
+    private final IgniteLogger log;
+
+    /** */
+    private final long time = U.currentTimeMillis();
+
+    /** */
+    public CollectPartitionStatistics(
+        LocalStatisticsGatheringContext gathCtx,
+        GridH2Table tbl,
+        Column[] cols,
+        int partId,
+        long ver,
+        IgniteLogger log
+    ) {
+        this.tbl = tbl;
+        this.cols = cols;
+        this.partId = partId;
+        this.ver = ver;
+        cancelled = () -> gathCtx.future().isCancelled();
+        this.log = log;
+    }
+
+    /** {@inheritDoc} */
+    @Override public ObjectPartitionStatisticsImpl call() {
+        CacheGroupContext grp = tbl.cacheContext().group();
+
+        GridDhtPartitionTopology top = grp.topology();
+        AffinityTopologyVersion topVer = top.readyTopologyVersion();
+
+        GridDhtLocalPartition locPart = top.localPartition(partId, topVer, false);
+
+        if (locPart == null)
+            return null;
+
+        boolean reserved = locPart.reserve();
+
+        try {
+            if (!reserved || (locPart.state() != OWNING)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Partition not owning. Need to retry [part=" + partId +
+                        ", tbl=" + tbl.identifier() + ']');
+                }
+
+                throw new GatherStatisticRetryException();
```

Comment:
With StatisticsGatherer::submitTask exception handling looks like the active wait.